src/docker/shared.rs

```diff
@@ -488,6 +489,271 @@ impl Directories {
     }
 }
 
+// the mount directory for the data volume.
+pub const MOUNT_PREFIX: &str = "/cross";
+// the prefix used when naming volumes
+pub const VOLUME_PREFIX: &str = "cross-";
+// default timeout to stop a container (in seconds)
+pub const DEFAULT_TIMEOUT: u32 = 2;
+// instant kill in case of a non-graceful exit
+pub const NO_TIMEOUT: u32 = 0;
+
+pub(crate) static mut CONTAINER: Container = Container::new();
+
+// we need to specify drops for the containers, but we
+// also need to ensure the drops are called on a
+// termination handler. we use an atomic bool to ensure
+// that the drop only gets called once, even if we have
+// the signal handle invoked multiple times or it fails.
+#[allow(missing_debug_implementations)]
+pub struct Container {
+    deleter: Option<DeleteContainer>,
+    exists: AtomicBool,
+}
+
+impl Container {
+    pub const fn new() -> Container {
+        Container {
+            deleter: None,
+            exists: AtomicBool::new(false),
+        }
+    }
+
+    pub fn create(engine: Engine, name: String) -> Result<()> {
+        // SAFETY: guarded by an atomic swap
+        unsafe {
+            if !CONTAINER.exists.swap(true, Ordering::SeqCst) {
+                CONTAINER.deleter = Some(DeleteContainer {
+                    engine,
+                    name,
+                    timeout: NO_TIMEOUT,
+                    color_choice: ColorChoice::Never,
+                    verbosity: Verbosity::Quiet,
+                });
+                Ok(())
+            } else {
+                eyre::bail!("attempted to create already existing container.");
+            }
+        }
+    }
+
+    // the static functions have been placed by the internal functions to
+    // verify the internal functions are wrapped in atomic load/stores.
+
+    pub fn exists(&self) -> bool {
+        self.exists.load(Ordering::Relaxed)
+    }
+
+    pub fn exists_static() -> bool {
+        // SAFETY: an atomic load.
+        unsafe { CONTAINER.exists() }
+    }
+
+    // when the `docker exec` command finished.
+    pub fn finish(&mut self, is_tty: bool, msg_info: &mut MessageInfo) {
+        // relax the no-timeout and lack of output
+        // ensure we have atomic ordering
+        if self.exists() {
+            let deleter = self.deleter.as_mut().unwrap();
+            if is_tty {
+                deleter.timeout = DEFAULT_TIMEOUT;
+            }
+            deleter.color_choice = msg_info.color_choice;
+            deleter.verbosity = msg_info.verbosity;
+        }
+
+        self.terminate();
+    }
+
+    pub fn finish_static(is_tty: bool, msg_info: &mut MessageInfo) {
+        // SAFETY: internally guarded by an atomic.
```

Comment:
```suggestion
        // SAFETY: internally guarded by an atomic swap.
```

Reply:
It's actually an atomic load.

Reply:
but the mechanic that makes it safe is that the operations are gated by a bool that is swapped, no?

Reply:
I guess a load and swap? The swap is for `terminate`. The load is for modifying the timeout and the verbosity/color choice.